src/Terminal.hs

```diff
@@ -0,0 +1,100 @@
+{-# LANGUAGE ImportQualifiedPost #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Terminal (terminalGame) where
+
+import Control.Applicative (liftA2)
+import Control.Monad ((<=<))
+import Data.Either.Extra (maybeToEither)
+import Data.Function (on)
+import Data.Map.Strict (empty, findWithDefault)
+import Data.Text qualified as Text
+import Data.Text.IO qualified as TextIO
+import GameLogic (Board (..), Coordinates, Player (..), nextTurn, playAtSpot, winner)
+import System.Console.ANSI (Color (..), ColorIntensity (..), ConsoleLayer (..), SGR (..), clearScreen, setCursorPosition, setSGR)
+import Text.Read (readMaybe)
+
+showPlayerT :: Player -> Text.Text
+showPlayerT = Text.pack . show
+
+showBoard :: Board -> Text.Text
+showBoard (Board board) =
+  "  0   1   2 \n"
+    <> Text.intercalate "\n ---+---+---\n" rows
+  where
+    rows = rowString <$> (zip [0 ..] $ replicate 3 [0 .. 2])
+    rowString :: (Int, [Int]) -> Text.Text
+    rowString (colIndex, row) =
+      (Text.pack . show) colIndex
+        <> " "
+        <> Text.intercalate " | " ((\t -> findWithDefault " " (t, colIndex) textBoard) <$> row)
```

Comment:
You might want to look at `keys` or `assocs` to make sure you can just `fold` (strictly).
The map is ordered, so you can count on folds to give you the right order of spots (you just need to group them into rows).
(Obviously more efficient: `O(size board)` vs `O(size board * log (size board))`).

Reply:
But I also need the keys that aren't in the board so `keys` and `assocs` don't really help me

Reply:
You don't want them anyways, because a custom fold is needed here.
A fold is equivalent to an imperative loop, so in the loop body you can see if you've skipped any missing coordinates between the previous one and the current one and deal with them as well as the current one.